_posts/2021-01-19-bazel-4-0.md

```diff
@@ -0,0 +1,99 @@
+* Bazel 4.0 is our first release with [long-term support](https://blog.bazel.build/2020/11/10/long-term-support-release.html).
+
+## Incompatible changes
+
+The following flags have been flipped:
+
+* [`--incompatible_blacklisted_protos_requires_proto_info`](https://github.com/bazelbuild/bazel/issues/11694)
+* [`--incompatible_force_strict_header_check_from_starlark`](https://github.com/bazelbuild/bazel/issues/11679)
+* [`incompatible_require_linker_input_cc_api`](https://github.com/bazelbuild/bazel/issues/10860): C++ Starlark API now requires `linker_inputs` wrapping `library_to_link`.
+* [`--incompatible_proto_output_v2`](https://github.com/bazelbuild/bazel/issues/10358) for aquery, which applies a new, more compact aquery proto output format.
+
+The following flag has been added:
+
+* [`--incompatible_display_source_file_location`](https://github.com/bazelbuild/bazel/issues/8900) has been added for `bazel query location=output` to print the line 1 location of the actual source files instead of the source file targets.
+
+The following flags have been enabled by default:
+
+* [`--incompatible_objc_compile_info_migration`](https://github.com/bazelbuild/bazel/issues/10854)
+* [`--incompatible_objc_provider_remove_compile_info`](https://github.com/bazelbuild/bazel/issues/11359)
+* [`--incompatible_run_shell_command_string`](https://github.com/bazelbuild/bazel/issues/5903)
+* [`--incompatible_restrict_string_escapes=true`](https://github.com/bazelbuild/bazel/issues/8380) . Unnecessary backslashes such as `\.` in string literals are now an error, instead of silently being treated as `\\.`. To fix the error while preserving behavior, manually add another backslash. However, the error often signals the original code needs to be updated.
+* [`--incompatible_avoid_conflict_dlls=true`](https://github.com/bazelbuild/bazel/issues/11600)
+
+The following flags have either been removed or are no-ops:
+
+* [`--incompatible_load_java_rules_from_bzl`](https://github.com/bazelbuild/bazel/issues/8746) and [`--incompatible_load_proto_rules_from_bzl`](https://github.com/bazelbuild/bazel/issues/8922) are now no-ops.
+* [`--incompatible_string_replace_count`](https://github.com/bazelbuild/bazel/issues/11244) has been flipped and removed.
+* `--local_resources` and `--incompatible_remove_local_resources` [have been removed](https://github.com/bazelbuild/bazel/issues/10536). If you've been setting `--local resources` or `--incompatible_remove_local_resources=false`, migrate to using `--local_ram_resources` and `--local_cpu_resources` instead.
+* The no-op [`--deep_execroot`](https://github.com/bazelbuild/bazel/issues/12271) flag has been removed.
+
+## User Interface
+
+* Error messages emitted when an action fails have been reworked to be more informative about the failing action.
+
+## Query
+
+* Querying with `output=location` now allows the `relative_locations` flag to properly display relative locations instead of the full path. Fixes #3497.
```

Comment:
Other GitHub issues are linked. Consider linking #3497 for consistency and ease.

Reply:
Done.